### PR TITLE
Dataflow annotations for APEX

### PR DESCRIPTION
**Pull request by biddisco**
Fixes #4257 (again) since the previous fixes were not complete.

### CODE REVIEW

File: hpx/lcos/dataflow.hpp [hunk 1]
```diff
@@ -59,8 +59,52 @@
...
+    template <typename Frame>
+    struct dataflow_finalization {
+        //
+        dataflow_finalization(Frame *df) : this_(df) {}
+        using is_void = typename Frame::is_void;
+        //
+        template <typename Futures>
+        void operator()(Futures&& futures)
+        {
+            return this_->execute(is_void{}, std::move(futures));
```
Comment (hkaiser) [inline hpx/lcos/dataflow.hpp]: This should be using `std::forward` instead of `std::move`.

File: hpx/lcos/dataflow.hpp [hunk 2]
```diff
@@ -59,8 +59,52 @@
...
+namespace hpx { namespace lcos { namespace detail
+{
+    template <typename Frame>
+    struct dataflow_finalization {
+        //
+        dataflow_finalization(Frame *df) : this_(df) {}
+        using is_void = typename Frame::is_void;
+        //
+        template <typename Futures>
+        void operator()(Futures&& futures)
```
Comment (hkaiser) [inline hpx/lcos/dataflow.hpp]: Can the operator be `const`?

File: hpx/lcos/dataflow.hpp [hunk 3]
```diff
@@ -59,8 +59,52 @@
...
+    template <typename Frame>
+    struct dataflow_finalization;
+}}}
+
+namespace hpx { namespace traits {
+#if defined(HPX_HAVE_THREAD_DESCRIPTION)
+    ///////////////////////////////////////////////////////////////////////////
+    // traits specialization to get annotation from dataflow_finalization
+    template <typename Frame>
+    struct get_function_annotation<lcos::detail::dataflow_finalization<Frame>>
```
Comment (hkaiser) [inline hpx/lcos/dataflow.hpp]: Please add `#include <hpx/functional/traits/get_function_annotation.hpp>`

File: hpx/lcos/local/packaged_continuation.hpp [hunk 4]
```diff
@@ -32,8 +32,49 @@
...
+    template <typename Continuation>
+    struct continuation_finalization;
+}}}
+
+namespace hpx { namespace traits {
+    #if defined(HPX_HAVE_THREAD_DESCRIPTION)
+        ///////////////////////////////////////////////////////////////////////////
+        // traits specialization to get annotation from dataflow_finalization
+        template <typename Continuation>
+        struct get_function_annotation<lcos::detail::continuation_finalization<Continuation>>
```
Comment (hkaiser) [inline hpx/lcos/local/packaged_continuation.hpp]: Please add `#include <hpx/functional/traits/get_function_annotation.hpp>`.

File: hpx/lcos/local/packaged_continuation.hpp [hunk 5]
```diff
@@ -376,8 +422,9 @@ namespace hpx { namespace lcos { namespace detail
             }
 
             hpx::intrusive_ptr<continuation> this_(this);
-            hpx::util::thread_description desc(f_,
-                "hpx::parallel::execution::parallel_executor::post");
+            char const* annotation = hpx::traits::get_function_annotation<
+                function_type>::call(f_);
```
Comment (hkaiser) [inline hpx/lcos/local/packaged_continuation.hpp]: What will be the fallback in this case (i.e. no specialization for `get_function_annotation` is found?

File: hpx/lcos/local/packaged_continuation.hpp [hunk 6]
```diff
@@ -32,8 +32,49 @@
...
+            static char const* call(
+                lcos::detail::continuation_finalization<Continuation> const& f) noexcept
+            {
+                char const* annotation = hpx::traits::get_function_annotation<
+                    typename hpx::util::decay<function_type>::type>::call(f.this_->f_);
+                return annotation;
+            }
+        };
+    #endif
+}}
```
Comment (hkaiser) [inline hpx/lcos/local/packaged_continuation.hpp]: Is this specialization really needed? The code below already calls `get_function_annotation` on the embedded function `f_`. Please explain.

File: hpx/lcos/dataflow.hpp (hunk 1 above)
Comment (biddisco) [reply hpx/lcos/dataflow.hpp]: yup. one of the builders failed and I found it too. I'll force push these commits shortly

File: hpx/lcos/local/packaged_continuation.hpp (hunk 6 above)
Comment (biddisco) [reply hpx/lcos/local/packaged_continuation.hpp]: The invocations for version with and without executors and different policies go through different places. I'm not sure I know which bits you are referring to as unneeded.

File: hpx/lcos/local/packaged_continuation.hpp (hunk 5 above)
Comment (biddisco) [reply hpx/lcos/local/packaged_continuation.hpp]: The default is to return "<unknown>" or something of that kind. I suppose I'll need to check and see what appears in the traces.

File: hpx/lcos/local/packaged_continuation.hpp (hunk 6 above)
Comment (hkaiser) [reply hpx/lcos/local/packaged_continuation.hpp]: Well, I think the specialization I attached this comment onto is not really needed... What is the use case (i.e. test) that is fixed by _this_ particular change?

Comment (biddisco) [reply hpx/lcos/local/packaged_continuation.hpp]:
```
    {
        hpx::future<int> f1 = hpx::make_ready_future(5);
        results.emplace_back(f1.then(
            NP_executor,
            hpx::util::annotated_function(
                [](auto &&f1) {
                    dummy_task(std::size_t(1000));
                },
                "Continuation with executor")));
    }
```

Comment (hkaiser) [reply hpx/lcos/local/packaged_continuation.hpp]: That code seems to work completely without your patch. I.e.
```
#include <hpx/hpx.hpp>
#include <hpx/hpx_main.hpp>

int main()
{
    hpx::future<int> f1 = hpx::make_ready_future(5);
     hpx::future<void> f2 = f1.then(hpx::parallel::execution::parallel_executor{},
        hpx::util::annotated_function(
            [](auto&& f1)
            {
                std::cout << hpx::threads::get_thread_description(
                    hpx::threads::get_self_id()) << "\n";
            }, "Continuation with executor"));
    f2.get();
    return 0;
}
```
Will print `Continuation with executor` without your patch. What do I miss?

BTW: could you add some tests for your changes, please?

Comment (biddisco) [reply hpx/lcos/local/packaged_continuation.hpp]: try it with 
```
    hpx::threads::scheduled_executor NP_executor =
           hpx::threads::executors::pool_executor(
               "default", hpx::threads::thread_priority_default);
  
```

Comment (biddisco) [reply hpx/lcos/local/packaged_continuation.hpp]: There isn't really a way of testing this. you have to use tools like vampir to load the generated otf files to see if the functions are visible. The file `tests/unit/resource/scheduler_priority_check.cpp` on my current branch has been extended with annotations for some of these and I will commit it to the branch once I'm happy that apex works. The last time I was able to use apex was back in early 2018, so there's no great rush with this.

Comment (biddisco) [reply hpx/lcos/local/packaged_continuation.hpp]: The tests I'm running are ...

```
    hpx::threads::scheduled_executor NP_executor =
        hpx::threads::executors::pool_executor(
            "default", hpx::threads::thread_priority_default);

    std::vector<hpx::future<void>> results;

    {
        hpx::future<int> f1 = hpx::make_ready_future(5);
        hpx::future<int> f2 = hpx::make_ready_future(5);
        results.emplace_back(hpx::dataflow(
            NP_executor,
            hpx::util::annotated_function(
                [](auto &&f1, auto &&f2) {
                    dummy_task(std::size_t(1000));
                },
                "Dataflow with executor"), f1, f2));
    }

    {
        hpx::future<int> f1 = hpx::make_ready_future(5);
        hpx::future<int> f2 = hpx::make_ready_future(5);
        results.emplace_back(hpx::dataflow(
            hpx::util::annotated_function(
                [](auto &&f1, auto &&f2) {
                    dummy_task(std::size_t(1000));
                },
                "Dataflow without executor"), f1, f2));
    }

    {
        hpx::future<int> f1 = hpx::make_ready_future(5);
        hpx::future<int> f2 = hpx::make_ready_future(5);
        results.emplace_back(hpx::dataflow(
            hpx::launch::sync,
            hpx::util::annotated_function(
                [](auto &&f1, auto &&f2) {
                    dummy_task(std::size_t(1000));
                },
                "Dataflow with sync policy"), f1, f2));
    }

    {
        hpx::future<int> f1 = hpx::make_ready_future(5);
        results.emplace_back(f1.then(
            NP_executor,
            hpx::util::annotated_function(
                [](auto &&f1) {
                    dummy_task(std::size_t(1000));
                },
                "Continuation with executor")));
    }

    {
        hpx::future<int> f1 = hpx::make_ready_future(5);
        results.emplace_back(f1.then(
            hpx::util::annotated_function(
                [](auto &&f1) {
                    dummy_task(std::size_t(1000));
                },
                "Continuation without executor")));
    }

    {
        hpx::future<int> f1 = hpx::make_ready_future(5);
        results.emplace_back(f1.then(
            hpx::launch::sync,
            hpx::util::annotated_function(
                [](auto &&f1) {
                    dummy_task(std::size_t(1000));
                },
                "Continuation with sync policy")));
    }

    // wait for completion
    hpx::when_all(results).get();
```

Comment (biddisco) [reply hpx/lcos/local/packaged_continuation.hpp]: I should add some other policies too ...

Comment (biddisco) [reply hpx/lcos/local/packaged_continuation.hpp]: I forgot to add the combinations that use `hpx:unwrapping` - I will commit a test shortly to the branch

**Comment by biddisco** [comment]
I added a check/test that can be run and visually inspected for errors

Without the fixes in this PR
```
Timer                                                : #calls  |    mean  |   total  |  % total  
------------------------------------------------------------------------------------------------
                                      - Continuation :        1      0.001      0.001      0.839
                - Executor non-threaded Continuation :        1      0.001      0.001      0.805
                                - async Continuation :        2      0.001      0.002      1.617
                                 - fork Continuation :        1      0.001      0.001      0.853
                                           <unknown> :        2      0.001      0.002      1.603
                        async_launch_policy_dispatch :        3      0.000      0.000      0.078
            broadcast_call_shutdown_functions_action :        2      0.000      0.000      0.163
                      call_shutdown_functions_action :        2      0.001      0.003      2.243
   hpx::parallel::execution::parallel_executor::post :        9      0.001      0.008      5.716
                              load_components_action :        1      0.006      0.006      4.746
                   primary_namespace_colocate_action :        2      0.000      0.000      0.105
                                          run_helper :        1      0.012      0.012      9.336
                                 shutdown_all_action :        1      0.001      0.001      0.512
                                           APEX Idle :                          0.094     71.386
------------------------------------------------------------------------------------------------
```
This PR
```
Timer                                                : #calls  |    mean  |   total  |  % total                                                                                               
------------------------------------------------------------------------------------------------                                                                                              
                                      - Continuation :        1      0.001      0.001      0.534                                                                                              
                                          - Dataflow :        1      0.001      0.001      0.531                                                                                              
                - Executor non-threaded     Dataflow :        1      0.001      0.001      0.553                                                                                              
                - Executor non-threaded Continuation :        1      0.001      0.001      0.521
                    - Executor threaded     Dataflow :        1      0.001      0.001      0.526
                    - Executor threaded Continuation :        1      0.001      0.001      0.524
                                - apply     Dataflow :        1      0.001      0.001      0.528
                                - async     Dataflow :        2      0.001      0.002      1.055
                                - async Continuation :        2      0.001      0.002      1.068
                             - deferred     Dataflow :        1      0.001      0.001      0.531
                                 - fork     Dataflow :        1      0.001      0.001      0.514
                                 - fork Continuation :        1      0.001      0.001      0.528
hpx::components::server::runtime_support::shutdow... :        1      0.001      0.001      0.294
hpx::components::server::runtime_support::load_co... :        1      0.016      0.016      7.879
hpx::components::server::runtime_support::call_sh... :        4      0.001      0.003      1.447
hpx::agas::server::primary_namespace::colocate_ac... :        2      0.000      0.000      0.072
hpx::actions::action<void (*)(hpx::components::se... :        2      0.000      0.000      0.109
                        async_launch_policy_dispatch :        3      0.000      0.000      0.089
                                          run_helper :        1      0.017      0.017      8.487
                                           APEX Idle :                          0.150     74.211
------------------------------------------------------------------------------------------------
```

**Comment by biddisco** [comment]
oops. had `async` in there twice. removed one and force pushed, counts in output will be off by one.

**Comment by biddisco** [comment]
note that I added a minus sin to the task names solely to make them appear at the start of the list alphabetically. I'll tweak that to make it more sensible next time.

**Comment by khuck** [comment]
@biddisco With `APEX_CSV_OUTPUT=1` you get the output above (or you can call `apex::apex_options::use_screen_output(true);` at the start of the test program).  Having done either of those two, you can add these types of test properties to the ctest definition:

```
set_tests_properties(dataflow_apex_test PROPERTIES PASS_REGULAR_EXPRESSION "- Dataflow :        1")
set_tests_properties(dataflow_apex_test PROPERTIES PASS_REGULAR_EXPRESSION "- Executor non-threaded     Dataflow :        1")
...
```

...or call `apex::get_profile` like happens in hpx/tests/unit/apex/apex_action_count.cpp and do the test programmatically.

**Comment by biddisco** [comment]
Thanks @khuck - I have added a call to `apex::apex_options::use_screen_output(true);` and prefixed the task names in such a way that they appear alphabetically in a checkable way. Then with ctest we can scan for the right string and pass/fail the test.

The git log includes this comment - even with the patches on my branch, not all tasks are correctly detected. There is still work to be done, but this makes the majority of tasks we are using in our cholesky code visible, (though not yet on N>1 mpi ranks).
```
    several continuation tasks are not correctly picked up by apex.
    hpx::launch::sync is not handled by apex
    hpx::unwrapping hides annotations - this should be addressed
```
I'm currently testing for the following task names (I'm not checking the count, what we really care about for now is if they stop appearing in the output). 
```
719: Timer                                                : #calls  |    mean  |   total  |  % total  
719: ------------------------------------------------------------------------------------------------
719:                                           1-Dataflow :        1      0.001      0.001      0.626
719:                     1-Executor non-threaded Dataflow :        1      0.001      0.001      0.601
719:                         1-Executor threaded Dataflow :        1      0.001      0.001      0.621
719:                                     1-apply Dataflow :        1      0.006      0.006      3.353
719:                                     1-async Dataflow :        1      0.001      0.001      0.612
719:                                  1-deferred Dataflow :        1      0.001      0.001      0.640
719:                                      1-fork Dataflow :        1      0.001      0.001      0.613
719:                                       2-Continuation :        1      0.001      0.001      0.619
719:                 2-Executor non-threaded Continuation :        1      0.001      0.001      0.608
719:                     2-Executor threaded Continuation :        1      0.001      0.001      0.621
719:                                 2-async Continuation :        1      0.001      0.001      0.625
719:                                  2-fork Continuation :        1      0.001      0.001      0.622
```
(continuations are still missing `apply`, `deferred` launch policies)

**Comment by biddisco** [comment]
Question : since `hpx::annotate_function();` the RAII alternative is no longer functioning (and hence things like `launch::sync` don't work) should we remove it altogether and all places that call it? There are still quite a lot of places in the code where it is sprinkled around.

**Comment by hkaiser** [comment]
> e code where it is sprinkled around.

Did it ever properly 'work' (for APEX)? Is there a way to make it work for APEX? @khuck?

Personally, I would prefer having a RAII version of the functionality. IIUC, today `hpx::annotate_function()` does not really annotate anything but forces a typetrait specialization to be generated that will cause anybody that asks for the name of a function of this function-type will get the annotated string returned. Is my understanding correct?

**Comment by khuck** [comment]
> Thanks @khuck - I have added a call to `apex::apex_options::use_screen_output(true);` and prefixed the task names in such a way that they appear alphabetically in a checkable way. Then with ctest we can scan for the right string and pass/fail the test.
> 
> The git log includes this comment - even with the patches on my branch, not all tasks are correctly detected. There is still work to be done, but this makes the majority of tasks we are using in our cholesky code visible, (though not yet on N>1 mpi ranks).
> 
> ```
>     several continuation tasks are not correctly picked up by apex.
>     hpx::launch::sync is not handled by apex
>     hpx::unwrapping hides annotations - this should be addressed
> ```
> 
> I'm currently testing for the following task names (I'm not checking the count, what we really care about for now is if they stop appearing in the output).
> 
> ```
> 719: Timer                                                : #calls  |    mean  |   total  |  % total  
> 719: ------------------------------------------------------------------------------------------------
> 719:                                           1-Dataflow :        1      0.001      0.001      0.626
> 719:                     1-Executor non-threaded Dataflow :        1      0.001      0.001      0.601
> 719:                         1-Executor threaded Dataflow :        1      0.001      0.001      0.621
> 719:                                     1-apply Dataflow :        1      0.006      0.006      3.353
> 719:                                     1-async Dataflow :        1      0.001      0.001      0.612
> 719:                                  1-deferred Dataflow :        1      0.001      0.001      0.640
> 719:                                      1-fork Dataflow :        1      0.001      0.001      0.613
> 719:                                       2-Continuation :        1      0.001      0.001      0.619
> 719:                 2-Executor non-threaded Continuation :        1      0.001      0.001      0.608
> 719:                     2-Executor threaded Continuation :        1      0.001      0.001      0.621
> 719:                                 2-async Continuation :        1      0.001      0.001      0.625
> 719:                                  2-fork Continuation :        1      0.001      0.001      0.622
> ```
> 
> (continuations are still missing `apply`, `deferred` launch policies)

The `sync` version doesn't create a task, so there is no APEX instrumentation around it.  Which is interesting, because direct actions are instrumented, if I remember correctly.

**Comment by msimberg** [comment]
I think the CircleCI failure is unrelated. I've restarted that build.

File: hpx/lcos/local/packaged_continuation.hpp (hunk 6 above)
Comment (hkaiser) [reply hpx/lcos/local/packaged_continuation.hpp]: I still believe, that this specialization is not really needed. Please verify. The code below already calls `get_function_annotation` on the embedded function `f_`. That should already be sufficient.

File: tests/unit/apex/annotation_check.cpp [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        auto now = std::chrono::steady_clock::now();
+        auto elapsed =
+            std::chrono::duration_cast<std::chrono::microseconds>(now - start);
+        sleep = (elapsed < std::chrono::microseconds(n));
+    } while (sleep);
+}
+
+// --------------------------------------------------------------------------
+// string for a policy
+std::string policy_string(const hpx::launch& policy)
```
Comment (hkaiser) [inline tests/unit/apex/annotation_check.cpp]: We normally use the east-const notation in our code base (not to start a religious war...). I think we should be consistent.

**Comment by biddisco** [comment]
> Why did you clang-format the whole file? This makes it close to impossible to review things.

Our checks run clang format on the whole file. If there is a way of just formatting the part you modified (and have it still pass our tests), please tell me.

**Comment by hkaiser** [comment]
> Why did you clang-format the whole file? This makes it close to impossible to review things.
> 
> Our checks run clang format on the whole file. If there is a way of just formatting the part you modified (and have it still pass our tests), please tell me.

Our checks run clang-format on all files in the modules (`libs`) directory only. There we do the whole-file-formatting as a step on its own (we're trying not to change things while moving a file into a module as much as possible). We have not formatted any of the other files (outside of `libs`) yet and I believe there is no reason to do so. 

So generally speaking, if we do format a whole file (for whatever reason), we do try to separate the formatting from the rest of the changes to enable reviewing.

**Comment by msimberg** [comment]
@biddisco `clang-format` also has an option to format specific lines only: `clang-format -lines=start:end`. It's quite horrible to use manually but many editors have an option to `clang-format` only a selection.

**Comment by aurianer** [comment]
There is also a `git-clang-format`used to format the diff of the commits, but I haven't tried it for now

**Comment by biddisco** [comment]
I removed most of the clang format stuff and squashed the tiny codacy and clang format changes into one commit.

**Comment by biddisco** [comment]
I removed the clan-formatted files and cleaned up inspect, but now clang-format test fails again
```
#!/bin/bash -eo pipefail
cd /hpx/source/libs && shopt -s globstar # to activate the ** globbing
clang-format-8 -i **/*.{cpp,hpp}
git diff --exit-code > /tmp/modified_format_files.txt

Exited with code exit status 1
```
It really looks to me as though it just clang formats all files in the commit and fails you if there is any non zero output. I waste far too much time on inspect and format errors.

**Comment by biddisco** [comment]
ok. my bad. it checks the files in lib and I made some changes there. Cleaned up again and force pushed. fingers crossed.

**Comment by msimberg** [comment]
@biddisco note that the end goal is that one can enable auto-format on save and you never have to think about formatting again. As soon as we get everything formatted once things should become much easier.

**Comment by kordejong** [comment]
I have tested commit 8707534 and can confirm that I now get useful information in my Vampir plots. When annotating lambdas passed to dataflow this information ends up in the trace. Great!

**Comment by msimberg** [comment]
What's currently blocking this?

**Comment by hkaiser** [comment]
> What's currently blocking this?

I'm still not convinced that the specializations added are actually needed. See my comments above.

**Comment by biddisco** [comment]
`hpx::parallel::execution::parallel_executor` isn't a thread_pool (or thread based?) executor and it goes through a different customization point I expect. This PR fixed my problems and the user's ones, but if it isn't of value then just delete it and I'll maintain my own branch for using apex.

**Comment by hkaiser** [comment]
@biddisco I have rebased this branch onto master. Let's see how the tests fare.

File: libs/execution/include/hpx/execution/executors/thread_pool_executor.hpp [hunk 8]
```diff
@@ -90,8 +90,9 @@ namespace hpx { namespace parallel { namespace execution {
             threads::thread_schedule_hint schedulehint, launch policy, F&& f,
             Ts&&... ts)
         {
-            hpx::util::thread_description desc(
-                f, "hpx::parallel::execution::parallel_executor::post");
+            char const* annotation =
+                hpx::traits::get_function_annotation<F>::call(f);
+            hpx::util::thread_description desc(f, annotation);
```
Comment (msimberg) [inline libs/execution/include/hpx/execution/executors/thread_pool_executor.hpp:95]: @biddisco was this change actually needed? None of the other places doing the same have changed... If I'm reading the constructor for `thread_description` correctly it will first try to use `hpx::traits::get_function_annotation<F>::call(f)`, and only if it's null will it use the fallback (the second argument). If we don't want the fallback that can be left out.

File: libs/execution/include/hpx/execution/executors/parallel_executor.hpp [hunk 9]
```diff
@@ -206,8 +206,9 @@ namespace hpx { namespace parallel { namespace execution {
         template <typename F, typename... Ts>
         void post(F&& f, Ts&&... ts) const
         {
-            hpx::util::thread_description desc(
-                f, "hpx::parallel::execution::parallel_executor::post");
+            char const* annotation =
+                hpx::traits::get_function_annotation<F>::call(f);
+            hpx::util::thread_description desc(f, annotation);
```
Comment (msimberg) [inline libs/execution/include/hpx/execution/executors/parallel_executor.hpp:211]: Same question as in `thread_pool_executor.hpp`.

**Comment by hkaiser** [comment]
@msimberg I have added more assertions here that make sure `local_thread_num` is not being used whenever it happens to be `-1`. Those asserts fire, unfortunately - looks like your fixes have not solved all the related problems.

@biddisco The guided executor is still failing to compile since this branch was rebased onto master. I don't know what is going on, care to have a look (https://app.circleci.com/pipelines/github/STEllAR-GROUP/hpx/2899/workflows/c2f4ac28-1268-4c02-a04b-19f0fd4fa8f6/jobs/177995)?

**Comment by msimberg** [comment]
> @msimberg I have added more assertions here that make sure `local_thread_num` is not being used whenever it happens to be `-1`. Those asserts fire, unfortunately (https://app.circleci.com/pipelines/github/STEllAR-GROUP/hpx/2899/workflows/c2f4ac28-1268-4c02-a04b-19f0fd4fa8f6/jobs/177989) - looks like your fixes have not solved all the related problems.

`local_num` is still allowed to be `-1`. I had intended for only `thread_num` to be used later which is not allowed to be `-1`. I've added asserts for `thread_num` instead.

I've merged master into this branch now (I've not fixed building the `async_customization` example).

File: libs/schedulers/include/hpx/schedulers/shared_priority_queue_scheduler.hpp [hunk 10]
```diff
@@ -275,10 +275,11 @@ namespace hpx { namespace threads { namespace policies {
 
                 // ------------------------------------------------------------
                 bool cleanup_terminated(
-                    std::size_t thread_num, bool delete_all) override
+                    std::size_t /*thread_num*/, bool delete_all) override
                 {
                     std::size_t local_num = local_thread_number();
-                    HPX_ASSERT(local_num < num_workers_);
+                    HPX_ASSERT(local_num != std::size_t(-1) &&
+                        local_num < num_workers_);
```
Comment (hkaiser) [inline libs/schedulers/include/hpx/schedulers/shared_priority_queue_scheduler.hpp:282]: Didn't you say `local_num` is _allowed_ to be `-1`?

File: libs/schedulers/include/hpx/schedulers/shared_priority_queue_scheduler.hpp [hunk 11]
```diff
@@ -406,12 +411,14 @@ namespace hpx { namespace threads { namespace policies {
                             d_lookup_[local_num] == domain_num)
```
Comment (hkaiser) [inline libs/schedulers/include/hpx/schedulers/shared_priority_queue_scheduler.hpp:411]: This operation will still fail if `local_num == -1`

**Comment by biddisco** [comment]
Changes to thread numbering have nothing to do with this dataflow-annotations branch. They should not be included here. If other changes have been made and merged into this, they should be fixed in their own PR's and branches.

File: libs/schedulers/include/hpx/schedulers/shared_priority_queue_scheduler.hpp [hunk 12]
```diff
@@ -618,7 +626,8 @@ namespace hpx { namespace threads { namespace policies {
                     bool enable_stealing) override
                 {
                     std::size_t this_thread = local_thread_number();
-                    HPX_ASSERT(this_thread < num_workers_);
+                    HPX_ASSERT(this_thread != std::size_t(-1) &&
+                        this_thread < num_workers_);
```
Comment (hkaiser) [inline libs/schedulers/include/hpx/schedulers/shared_priority_queue_scheduler.hpp:630]: Same here, this will fire if `thread_num == -1`.

**Comment by hkaiser** [comment]
> Changes to thread numbering have nothing to do with this dataflow-annotations branch. They should not be included here. If other changes have been made and merged into this, they should be fixed in their own PR's and branches.

Right... The only problem is that the issues manifest themselves on this branch only, I think.

File: libs/schedulers/include/hpx/schedulers/shared_priority_queue_scheduler.hpp (hunk 10 above)
Comment (msimberg) [reply libs/schedulers/include/hpx/schedulers/shared_priority_queue_scheduler.hpp:282]: Ok, almost always... I didn't want to change the behaviour here, just make it clear that `thread_num` is not used. And my changed assert doesn't make much sense with `std::size_t`. In any case, @biddisco will drop these changes and redo the checks if necessary.

**Comment by biddisco** [comment]
I will fix this branch and the thread numbering.

**Comment by msimberg** [comment]
@biddisco are you able to reproduce the segfaults with the apex builder? They might be the failures we were trying to fix earlier.

**Comment by biddisco** [comment]
I do not know what segfaults you are referring to. The dataflow PR looks pretty clean on cdash to me - could you please elaborate.

**Comment by msimberg** [comment]
https://cdash.cscs.ch//viewTest.php?onlyfailed&buildid=105728

**Comment by biddisco** [comment]
Those tests appear to be failing on multiple dashboards 
https://cdash.cscs.ch//testSummary.php?project=2&name=tests.performance.local.resume_suspend&date=2020-04-27
it does not appear to be connected to this PR

**Comment by msimberg** [comment]
> Those tests appear to be failing on multiple dashboards
> https://cdash.cscs.ch//testSummary.php?project=2&name=tests.performance.local.resume_suspend&date=2020-04-27
> it does not appear to be connected to this PR

The tests failed on those two PRs (#4570 and #4551) because the build failed (no tests to run). The tests are not failing on master.

**Comment by biddisco** [comment]
OK. I'll delete this PR and come back to it some other time.

**Comment by biddisco** [comment]
Reopening this after debugging with a clang build. No idea why gcc builds pass and clang ones fail, because they were both dereferencing a nullptr that was taken from the default implementation of `get_function_annotation` - this is fixed by returning "<anonymous>" instead of nullptr so that apex always has something to store. This might fix other apex related problems that I was having when reductions of task names took place in distributed runs. I will check this. 
Thanks @msimberg for being annoying enough that I took some time to investigate this.

**Comment by biddisco** [comment]
FYI: rebased onto latest master.

**Comment by msimberg** [comment]
> Thanks @msimberg for being annoying enough that I took some time to investigate this.

You're welcome, I guess... how come the nullptrs aren't being dereferenced on master? Are there maybe some checks missing for nullptrs (the fallback strings are fine by me though, I'm just wondering what changed)?

**Comment by hkaiser** [comment]
> Reopening this after debugging with a clang build. No idea why gcc builds pass and clang ones fail, because they were both dereferencing a nullptr that was taken from the default implementation of get_function_annotation - this is fixed by returning "" instead of nullptr so that apex always has something to store. This might fix other apex related problems that I was having when reductions of task names took place in distributed runs. I will check this.

May I suggest using something more descriptive as the fallback - perhaps `"<unknown>"`? That might help identifying issues down the road.

Also, @biddisco thanks for not dropping the ball on this!

**Comment by biddisco** [comment]
> May I suggest using something more descriptive as the fallback - perhaps `"<unknown>"`? That might help identifying issues down the road.

It actually returns `"<anonymous>"` but I'm not sure why the markup cha nged my text to an empty string.

**Comment by biddisco** [comment]
> how come the nullptrs aren't being dereferenced on master? Are there maybe some checks missing for nullptrs (the fallback strings are fine by me though, I'm just wondering what changed)?

As far as I can tell, the nullptrs get derefernced to some random stack memory on gcc and cause a segfault in clang. I added printfs and gcc was dumping out rubbish, consistent with some of the stuff I saw being generated in OTF files that woulnd't load properly in vampir.

I should investigate further, but when I saw the nullptr stuff being accessed I changed the unspecialized version of `get_annotated_function` and everything started working, so I stopped then.

**Comment by msimberg** [comment]
@biddisco this just needs formatting now. I think we can merge this after that.

**Comment by msimberg** [comment]
@biddisco I just rebased and clang-formatted this branch. If CI comes back happy I'll merge it.
